### ISSUE TITLE
[annoucement] current state of Filestash

### ISSUE THREAD

**Issue by mickael-kerjean** [mickael-kerjean/filestash#629]
Hello everyone,

I just want to make a status update on the state of Filestash, what's coming in and why I haven't been able to be as responsive as before in answering everyone's questions.

The gist of it is I'm rewriting the frontend of Filestash from React to Vanilla JS. The goal of this rewrite is many fold:

1. fix all the frontend hacks, improve the quality. When I got started with Filestash in 2017, I had no idea what it would become and how things will be. It was not supposed to be a project many thousands of people rely on all around the world. There's a few things like tagging, having a sidebar navigation and enable embedding in your website that aren't simple to implement looking at the current structure of things, cleaning everything up will align the codebase to where Filestash is going in the long run.

2. fix all the maintenance issues and unwanted dependencies. JS tooling is a pain to follow with things changing on a daily/weekly basis, making the maintenance aspect a lot more challenging that it should, the simplest cra web app has > 800 hundreds dependencies. Our industry has accepted this fate but I think this is insanity, I want to be able to audit all my dependencies making sure everything is clean and I don't ever want to go through the pain of migrating a webpack 2.2 code base to webpack v5. A couple months ago, I started working on what a simple es6 spa application could look like with "modern" code using only standards. I came up with this https://mickael-kerjean.github.io/skeleton/example/index.html which is just a prototype to fix some ideas before starting on the migration. Things are moving quite fast, the plan is to first migrate the admin side of things, release it to the public, get some feedback then start the more visible frontoffice side

3. make it simple to extend the core so you can customise things for your needs. The objective here is to end up such as you can open your browser developer console and say "wow, this is so simple, I can understand everything, let's make a patch that will add the behaviour I want" where your patch is just a diff on what you want to change, all in real time, no compilation necessary with a very small barrier to entry to start hacking

4. regaining control over every single details of how the application work, what I want is the best possible performance on the world shittiest devices, and to do that we need absolute control of every single piece of js that is being loaded, from where, when and how. I've seen people with insane setup trying to load buckets with millions of files, and this is where things become tricky. Filestash should accomodate the craziest use cases, getting the best possible performance with 0 overhead and no unwanted js code running on your device so we can tailor things to the finest details for the platform you use it from. Framework are great but all those abstractions have a cost and it's time to peel all those convenience layers off as introducing new libraries is just not gonna solve those pain points

**Comment by neingeist** [comment]
Thanks for your efforts!

**Comment by mickael-kerjean** [comment]
I finally released the very first part of this migration which is effectively a rewrite of the admin console before tackling the actual end user frontend later on.

Some interesting numbers with this delivery: 

Build size:

|          | With brotli compression     | uncompressed | 
|--------------|-----------|------------|
| current vanilla JS | 60.6kB     | 156kB        |
|  previous react app     | 287kB | 2.8MB       |

There's some more opportunities to reduce this further but we're running effectively 95% less code than before on the browser.

Time to interaction with the admin console with a server that is 10,000km away from me and a shitty internet connection:
- react app with a build: 4.88s
- current solution without a build: 5.3s

Without introducing a build step and using plain HTTP1, we're running very close to the original admin console app. There's a few opportunities to make this much better than before:
- leverage the interesting part of HTTP2 and/or HTTP3
- introduce an optional build step if there's some additional gain to be taken from a pure HTTP2/3 solution

**Comment by sevmonster** [comment]
This is amazing. In today's dependency hell clown world, it's incredibly refreshing to see someone working toward a more sane benchmark instead of digging themselves deeper. Good work.

**Comment by mickael-kerjean** [comment]
The new frontend is out. You can use it by setting the env variable CANARY=true which was enabled via 3cd12f863f1650f85e31b41a1e2bb1bc8b9aee93 
The plan is to wait for bugs to be reported before removing the old frontend code and finalise a couple minor features that haven't been ported yet (most notably chromecast support)

**Comment by poVoq** [comment]
I played around with the canary a bit, and it all seems to work fine except the ONLYOFFICE integration.

**Comment by ElderBlade** [comment]
Great work! Is the sidebar new? After clicking the sidebar icon (Your Files), the sidebar closes and there's no way to ever open it again.

**Comment by Folas1337** [comment]
The search bar doesn't seem to overlay the sidebar by default so you can't see what you're typing. Something to fix along the way :)

**Comment by mickael-kerjean** [comment]
Thank you all. I've open some tickets so as to not get sidetracked with the details of all those:
@poVoq => #769
@ElderBlade => #770
@Folas1337 => #771 

If somebody find any other issue, feel free to make some reports either here, as a separate ticket or both

**Comment by mickael-kerjean** [comment]
I've introduced a new optimization that reduce the loading time by 2 with the new frontend. At this stage, we're by every metric possible better than the original React based frontend.

**Comment by steveiliop56** [comment]
Responsivness is really good for the most part. The admin console has some issues though ;)

<img width="264" alt="Image" src="https://github.com/user-attachments/assets/203c374a-a931-4623-b0c4-7beb0cb4b85d" />

Just the sidebar is broken, the actual settings are all super responsive.

**Comment by badguy001** [comment]
compile from commit [2c09815c4e5f6bd681a3a5cf88b24774848f2da5](https://github.com/mickael-kerjean/filestash/commit/2c09815c4e5f6bd681a3a5cf88b24774848f2da5). wopi plugin not works when React Developer Tools of brower plugin is enable. but using LEGACY=true, wopi plugin works. error in browser is:
```
type:    Error
message: "[object Object]" is not valid JSON
trace:   SyntaxError: "[object Object]" is not valid JSON
    at JSON.parse (<anonymous>)
    at http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/pages/viewerpage/application_iframe.js:22:34
    at http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:25057
    at q._next (http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:5728)
    at q.next (http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:2785)
    at http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:29760
    at q._next (http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:5728)
    at q.next (http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:2785)
    at t (http://acrh17-gz.v6.rocks:60123/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:28365)
```
event.data is:
```
{source: 'react-devtools-content-script', hello: true}
[[Prototype]]: Object
```

**Comment by MichalLauer** [comment]
So after a long night, I have figured out that the error

```javascript
type:    Error
message: "[object Object]" is not valid JSON
trace:   SyntaxError: "[object Object]" is not valid JSON
    at JSON.parse (<anonymous>)
    at http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/pages/viewerpage/application_iframe.js:22:34
    at http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:25057
    at q._next (http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:5728)
    at q.next (http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:2785)
    at http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:29760
    at q._next (http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:5728)
    at q.next (http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:2785)
    at t (http://xxx/assets/2c09815c4e5f6bd681a3a5cf88b24774848f2da5/lib/vendor/rxjs/rxjs-shared.min.js:3:28365)
```

is caused by grammarly. For some reason, after the document is fully loaded, this line of coded is triggered

https://github.com/mickael-kerjean/filestash/blob/dda17d0553b8468eaad5b45645e20f0645dedf31/server/plugin/plg_editor_wopi/handler.go#L167-L185

where `event.data` is `NULL` and it fails. After disabling Grammarly, it works like a charm! Hope this helps others.
